src/sql-parser/src/ast/defs/statement.rs

```diff
@@ -497,23 +500,104 @@ impl<T: AstInfo> AstDisplay for CreateSourceStatement<T> {
             f.write_node(&display::comma_separated(&self.include_metadata));
         }
 
-        match &self.envelope {
-            None => (),
-            Some(envelope) => {
-                f.write_str(" ENVELOPE ");
-                f.write_node(envelope);
-            }
+        if let Some(envelope) = &self.envelope {
+            f.write_str(" ENVELOPE ");
+            f.write_node(envelope);
         }
 
         if !self.with_options.is_empty() {
             f.write_str(" WITH (");
             f.write_node(&display::comma_separated(&self.with_options));
             f.write_str(")");
         }
+
+        if let Some(subsources) = &self.subsources {
+            f.write_str(" ");
+            f.write_node(subsources);
+        }
     }
 }
 impl_display_t!(CreateSourceStatement);
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// A selected subsource in a FOR TABLES (..) statement
+pub enum CreateSourceSubsource<T: AstInfo> {
+    /// An unaliased subbsource
+    Bare(UnresolvedObjectName),
+    /// An subbsource aliased to a different name
+    Aliased(UnresolvedObjectName, UnresolvedObjectName),
+    /// An subbsource fully resolved to the target catalog object it will be written to
+    Targeted(UnresolvedObjectName, T::ObjectName),
+}
+
+impl<T: AstInfo> AstDisplay for CreateSourceSubsource<T> {
+    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
+        match self {
+            Self::Bare(name) => f.write_node(name),
+            Self::Aliased(name, alias) => {
+                f.write_node(name);
+                f.write_str(" AS ");
+                f.write_node(alias);
+            }
+            Self::Targeted(name, target) => {
+                f.write_node(name);
+                f.write_str(" INTO ");
+                f.write_node(target);
+            }
+        }
+    }
+}
+impl_display_t!(CreateSourceSubsource);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum CreateSourceSubsources<T: AstInfo> {
+    /// A subset defined with FOR TABLES (...)
+    Subset(Vec<CreateSourceSubsource<T>>),
+    /// FOR ALL TABLES
+    All,
+}
+
+impl<T: AstInfo> AstDisplay for CreateSourceSubsources<T> {
+    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
+        match self {
+            Self::Subset(subsources) => {
+                f.write_str("FOR TABLES (");
+                f.write_node(&display::comma_separated(subsources));
+                f.write_str(")");
+            }
+            Self::All => f.write_str("FOR ALL TABLES"),
+        }
+    }
+}
+impl_display_t!(CreateSourceSubsources);
+
+/// `CREATE SUBSOURCE`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CreateSubsourceStatement<T: AstInfo> {
```

Comment:
Yeah I think we'd have to create a system table for that because the subsources are created before the source that is writting to them and so they can't have a `FROM SOURCE` reference, the AST name resolution wouldn't work. I'd like to leave that to a future PR if that's ok

Reply:
System table sgtm